### ISSUE TITLE
`cast sig` does not behave as expected on some inputs

### ISSUE THREAD

**Issue by zk-tarts** [foundry-rs/foundry#1521]
### Component

Cast

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

nightly-8478c0841b00789c735df68ad111105fa9204f55

### What command(s) is the bug in?

cast sig

### This should fail but does not
```sh 
cast sig "foo(uint999)"
0x54511ad7
```
### This should succeed
```sh 
cast sig "foo(function() external)"
# panic with 'Illegal abi' message
```
The root of the  problem is really over in ethers-rs. The parser does not handle function types in human readable ABI and does not make sure the bytes/int/uint sizes are correct, among other things.   ref https://github.com/gakonst/ethers-rs/issues/474

#### Arrays with unit sizes are *techincally* valid solidity
In solidity this expands out to `uint[60]`.
```solidity
foo(uint[1 minutes] memory)
```
#### Arrays have a maximum size in solidity
This gets parsed as correct but is not.
```sh
cast sig "foo(uint[99999999999999999])"
0xfece5cb9
```

**Comment by zk-tarts** [comment]
Should I have opened this issue in https://github.com/gakonst/ethers-rs/ ? It's really more of a problem there than it is a foundry bug.

**Comment by onbjerg** [comment]
>cast sig "foo(function() external)"

I don't think it should? You cannot make these functions external, so they will never have a 4byte signature.

The others make sense to me.

There is another issue: Some Cast commands (`cast call` for example), we cannot use structs. `AbiParser` supports it, but it really just expands them into tuples. We should probably allow people to pass tuples (even though it is technically invalid) to support people calling functions with structs (and functions that return structs)

**Comment by onbjerg** [comment]
@mattsse Assigning you to this since it's the same underlying issue as #2025 and you assigned yourself to that one

**Comment by mattsse** [comment]
function types are currently not supported by `ethabi`, there's simply not a type variant for them

tracked here https://github.com/rust-ethereum/ethabi/issues/273

**Comment by zk-tarts** [comment]
If ethabi doesn't add function types then you could just treat them as `bytes24`  https://docs.soliditylang.org/en/latest/abi-spec.html#types

**Comment by mattsse** [comment]
there are essentially two issues:
* encoding/decoding
* signature

which is handled here 

https://github.com/rust-ethereum/ethabi/blob/451d029c69a1fce7c4d57cc117e985a31dc53627/ethabi/src/param_type/writer.rs#L25-L52

and would require another variant

**Comment by onbjerg** [comment]
Should be fixed in #2151, please let me know if it's not!
